crates/python/src/execution_data.rs

```diff
@@ -153,9 +159,109 @@ impl PyRegisterMatrix {
 
 #[pymethods]
 impl PyRegisterMap {
-    pub fn get_register_matrix(&self, register_name: String) -> Option<PyRegisterMatrix> {
+    pub fn get_register_matrix(&self, register_name: &str) -> Option<PyRegisterMatrix> {
         self.as_inner()
-            .get_register_matrix(&register_name)
+            .get_register_matrix(register_name)
             .map(PyRegisterMatrix::from)
     }
+
+    pub fn __len__(&self) -> usize {
+        self.as_inner().0.len()
+    }
+
+    pub fn __contains__(&self, key: String) -> bool {
+        self.as_inner().0.contains_key(&key)
+    }
+
+    pub fn __getitem__(&self, item: &str) -> PyResult<PyRegisterMatrix> {
+        self.get_register_matrix(item)
+            .ok_or(PyKeyError::new_err(format!(
+                "Key {item} not found in RegisterMap"
+            )))
+    }
+
+    pub fn __iter__(&self, py: Python<'_>) -> PyResult<Py<PyRegisterMapKeysIter>> {
+        Py::new(
+            py,
+            PyRegisterMapKeysIter {
+                inner: self.as_inner().0.clone().into_iter(),
```

Comment:
Curious if it would be any different to do this:

```suggestion
                inner: self.as_inner().0.iter().cloned(),
```

Reply:
Also, shouldn't this be just the keys?

```suggestion
                inner: self.as_inner().0.keys().cloned(),
```

Reply:
Alternatively, why not just use the [pyo3 provided](https://docs.rs/pyo3/latest/pyo3/types/struct.PyIterator.html) `PyIterator`? Then you wouldn't need to define the various iterators below.

Reply:
Maybe I'm missing something, but I think `PyIterator` is just the Rust native type for interacting with Python iterators. That is, if you call some Python code from Rust that returns an iterator, a `PyIterator` is what you would want to put it in to work with it. I see no way to build or return one from a Rust iterator. This setup is based on  the [docs on iterable objects in pyo3](https://pyo3.rs/v0.19.2/class/protocols.html?highlight=iter#iterable-objects).

The problem with using `cloned()`, `keys()`, and `values()` is that they are all an iterator type based on a reference to the `HashMap`, which means the iterator struct would require lifetimes, which isn't compatible with `#[pyclass]`.

I'll add a comment to the iterator structs to help clarify.